File: Divide and Conquer/Counting Sort/code_edu.cpp
#include "algorithm-visualizer.h"
#include <iostream>

const int				ARRAYSIZE = /*(배열 길이 입력)*/
const int				max = /*(0~max 사이의 난수로 배열 초기화-max 값 입력)*/
//visualizer{
Array1DTracer			arrayTracer("Array");
Array1DTracer			countsTracer("Counts");
Array1DTracer			sortedArrayTracer("Sorted Array");
Randomize::Integer		integer(0, max);
Randomize::Array1D<int> array1D(ARRAYSIZE, integer);
int						_array[ARRAYSIZE];
int						counts[max + 1];

void InitCountingSort()
{
	Layout::setRoot(VerticalLayout({ arrayTracer,countsTracer,sortedArrayTracer }));
	array1D.fill(_array);
	arrayTracer.set(_array);
	countsTracer.set(counts);
	Tracer::delay();
	for (int i = 0; i < max; i++)
		counts[i] = 0;
}
//}

int main()
{
	//_array		: 정렬하고자 하는 배열
	//counts		: _array의 각 원소의 갯수를 저장할 배열
	//sortedArray	: 정렬된 _array가 들어갈 배열
	InitCountingSort();
	for (int i = 0; i < ARRAYSIZE; i++) {
		int number = /*(알맞은 값 입력)*/
		counts[number]++;
		//visualizer{
		arrayTracer.select(i);
		countsTracer.patch(number, counts[number]);
		Tracer::delay();
		countsTracer.depatch(number);
		arrayTracer.deselect(i);
		//}
	}

	for (int i = 1; i <= max; i++) {
		counts[i] += /*(알맞은 값 입력)*/

		//visualizer{
		countsTracer.select(i - 1);
		countsTracer.patch(i, counts[i]);
		Tracer::delay();
		countsTracer.depatch(i);
		countsTracer.deselect(i - 1);
		//}
	}

	int sortedArray[ARRAYSIZE];
	for (int i = 0; i < ARRAYSIZE; i++)
		sortedArray[i] = NULL;
	//visualizer{
	sortedArrayTracer.set(sortedArray);
	//}

	for (int i = ARRAYSIZE - 1; i >= 0; i--) {
		int number = _array[i];
		int count = counts[number];
		sortedArray[count - 1] = number;

		//visualizer{
		arrayTracer.select(i);
		countsTracer.select(number);
		sortedArrayTracer.patch(count - 1, sortedArray[count - 1]);
		Tracer::delay();
		sortedArrayTracer.depatch(count - 1);
		countsTracer.deselect(number);
		arrayTracer.deselect(i);
		//}

		/*(counts[number]값을 알맞게 변경)*/
	}
}

File: Divide and Conquer/Counting Sort/code_edu.java
import org.algorithm_visualizer.*;
import java.util.Arrays;

class Main {
    private static int sizeOfArray = /*(배열 크기 입력)*/
    private static int minValue = 0; /*(배열 최소값 입력)*/
    private static int maxValue = 9; /*(배열 최대값 입력)*/ 
    //visualizer {
	private static Array1DTracer arrayTracer = new Array1DTracer("Array");
	private static Array1DTracer countsTracer = new Array1DTracer("Counts");
	private static Array1DTracer sortedArrayTracer = new Array1DTracer("Sorted Array");
	private static Integer[] array = (Integer[]) new Randomize.Array1D(arraySize,
			new Randomize.Integer(minValue, maxValue)).create();
	//}
	private static int[] counts;
	private static int[] sortedArray;
  //private static int[] array;

	Main() {
	    //visualizer {
	    Layout.setRoot(new VerticalLayout(new Commander[] {arrayTracer,countsTracer,sortedArrayTracer }));
	    //}
		// array의 최댓값을 찾아, 최댓값의 크기를 이용하여 원소 개수를 저장할 count 배열 생성
		int max = array[0];
		for(int i =1; i < array.length; i++)
			if(array[i] > max) max = array[i];
		
		counts = new int[max+1];
		Arrays.fill(counts, 0);
		// visualize {
		arrayTracer.set(array);
		countsTracer.set(counts);
		Tracer.delay();
		// }

		// 각 수를 count 에저장
		for (int i = 0; i < array.length; i++) {
			int number = array[i];
			/*(각 수를 저장하는 식 입력)*/
			// visualize {
			arrayTracer.select(i);
			countsTracer.patch(number, counts[number]);
			Tracer.delay();
			countsTracer.depatch(number);
			arrayTracer.deselect(i);
			// }
		}

		// 누적 값 저장
		for (int i = 1; i <= max; i++) {
			/*(count 누적을 구하는 식 입력)*/
			// visualize {
			countsTracer.select(i - 1);
			countsTracer.patch(i, counts[i]);
			Tracer.delay();
			countsTracer.depatch(i);
			countsTracer.deselect(i - 1);
			// }
		}

		// create a sorted array based on the prefix sums
		sortedArray = new int[array.length];
		// visualize {
		sortedArrayTracer.set(sortedArray);
		// }
		for (int i = array.length - 1; i > 0; i--) {
			int number = array[i];
			int count = counts[/*(?)*/];
			sortedArray[count - 1] = /*(?)*/;
			// visualize {
			arrayTracer.select(i);
			countsTracer.select(number);
			sortedArrayTracer.patch(count - 1, sortedArray[count - 1]);
			Tracer.delay();
			sortedArrayTracer.depatch(count - 1);
			countsTracer.deselect(number);
			arrayTracer.deselect(i);
			// }
			counts[/*(?)*/]--;
		}
	}

	public static void main(String[] args) {
		new Main();
	}
}